docs/Configuration.md
# GUI Config

 Deskflow will automatically figure out where to save settings and other files.

## Search paths

Deskflow will look for settings in several places depending on your operating system.
The search order for a setting file depends on your operating system

### Linux

  1. `<XDG_CONFIG_HOME>/Deskflow/Deskflow.conf`
  2. `~/.config/Deskflow/Deskflow.conf`
  3. `/etc/Deskflow/Deskflow.conf`
 
A new settings file will be created in the user path if no settings file is found.
The path of the settings file will be used as the base for all other config files.

### macOS
 
  1. `~/Library/Deskflow/Deskflow.conf`
  2. `/Library/Deskflow/Deskflow.conf`
 
A new settings file will be created in the user path if no settings file is found.
The path of the settings file will be used as the base for all other config files.

### Windows

  1. `<install-path>/settings/Deskflow.conf`
  2. Windows Registry `HKCU\Software\Deskflow\Deskflow`

Windows will save to the install dir if settings are loaded from there. If not, it saves any other config files in: `C:\ProgramData\Deskflow\`

When using settings from the install dir, the service mode will not be available.

## Valid GUI Keys

The GUI config file contains several sections.
Each section is formatted the same.
Option-value pairs are only written if the value is not the default value.

```
[section]
option=value
```

### Client

This section contains options used when in client mode. 
It will begin with `[client]`

| Option                    |    Valid Values    | Description |
|:--------------------------|:------------------:|:-----------|
| dynamicConnectionInterval | `true` or `false`  | Use dynamic connection retry times based on number of previously failed attempts [default: false] |
| languageSync              | `true` or `false`  | Sync to server language [default: true] |
| remoteHost                | `IP` or `hostname` | The remote host(s) to connect to. Use a comma separated list when you want to try several hosts |
| yScrollScale              | Double 0.1 - 10.0  | Vertical mouse scrolling will be scaled by this amount on the client [default: 1.0] |
| xScrollScale              | Double 0.1 - 10.0  | Horizontal mouse scrolling will be scaled by this amount on the client [default: 1.0] |
| invertYScroll             | `true` or `false`  | Invert vertical scroll on this client [default: false] |
| invertXScroll             | `true` or `false`  | Invert horizontal scroll on this client [default: false] |

### Core

This section contains general options it will begin with `[core]`

|Option         | Valid Values|Description|
|:--------------|:-----------:|:-----------|
| coreMode      | `0` or `1` or `2` | The mode to start in 0: None, 1: Client, 2: Server [default: 0]|
| display       |  int              | The XWindow display to use [default: autodetected] |
| interface     | IP Address        | Preferred IP to use for network communication. By default the server board casts on any available address |
| lastVersion   | M.m.p.t           | The version last run used for checking for updates |
| port          | port #            | Port to use when connecting [default: 24800 |
| preventSleep  | `true` or `false` | Prevent sleep when Deskflow is active [default: false] |
| processMode   | `1` or `0`        | The mode we use to start the process Service or Desktop |
| computerName  | string            | Name used to identify the computer [default: machine's hostname] |
| useHooks      | `true` or `false` | If Windows uses hooks or not [default: true] |
| language      | 639 language      | The language to display the GUI in [default: en] |
| enableEnterCommand | `true` or `false` | Should the enter command be triggered when the computer is entered [defaut: false] |
| enterCommand  | command | A command to run when the computer is entered. |
| enableExitCommand | `true` or `false` | Should the exit command be triggered when the computer is exited [defaut: false] |
| exitCommand  | command | A command to run when the computer is exited. |

### Daemon

This section contains options used by the daemon on windows it will begin with `[daemon]`

|Option | Valid Values|Description|
|:----------|:-----------:|:-----------|
| command   | Filename          | The filename of the binary the daemon. This binary exists in the same path as the deskflow GUI |
| elevate   | `true` or `false` | Elevate the daemon app [default: true unless portable mode ] |
| logFile   | Filepath          | Filepath of the daemon log |
| logLevel  | valid log Level,  | Log Level  |

### GUI

This section contains options used by the GUI it will begin with `[gui]`

|Option                          | Valid Values      |Description|
|:-------------------------------|:-----------------:|:-----------|
| autoHide                       | `true` or `false` | When true the app will hide itself on start up [default: false] |
| enableUpdateCheck              | `true` or `false` | When true check the update URL to see if a new version was released on start up [default: false] |
| closeReminder                  | `true` or `false` | Used to track if we have shown the reminder that when you close the app it remain running in the background  [default: true]|
| closeToTray                    | `true` or `false` | When `true` the gui will run in the systemTray when its closed [default: true] |
| logExpanded                    | `true` or `false` | Should the log section of the GUI be opened [default: false] |
| symbolicTrayIcon               | `true` or `false` | When true use the monocolor (symbolic) icon false uses a colorful icon for the tray [default: true] |
| showGenericClientFailureDialog | `true` or `false` | When `true` client connection errors will not show popup error messages [default: true] |
| shownFirstConnectedMessage     | `true` or `false` | When `true` GUI has shown the user the message for connecting the first time [default: false] |
| shownServerFirstStartMessage   | `true` or `false` | When `true` GUI has shown the user the Deskflow server is now running message [default: false] |
| shownVerionInTitle             | `true` or `false` | When `true` GUI will include the version in the window title [default: false] |
| startCoreWithGui               | `true` or `false` | When true the Core will be started with the GUI. It is set to the Core's state on exit. |
| updateCheckUrl                 | URL               | The URL to use when checking for a new version number, it should return a version [default: https://api.deskflow.org/version]|

### Log

This section contains options used by the application logging it will begin with `[log]`

|Option    |    Valid Values   |Description|
|:---------|:-----------------:|:-----------|
| file     | Filepath          | The file to write the log into |
| level    | Valid log level   | Log level to use |
| toFile   | `true` or `false` | When true the log will be written to the value of the `file` option |
| guiDebug | `true` or `false` | When true the log will show the Gui's internal debug messages |

### Security

This section contains options used by the application security it will begin with `[security]`

|Option                 |   Valid Values    |Description|
|:----------------------|:-----------------:|:-----------|
| checkPeerFingerprints | `true` or `false` | When true peers will have their fingerprints confirmed by the user and stored [default: true] |
| certificate           | Filepath          | Path to the certificate to use to encrypt messages.|
| keySize               | `2048` OR `4096`  | Size of the TLS key to use [default: 2048]| 
| tlsEnabled            | `true` or `false` | Are we using TLS encryption when communicating [default: true].|

### Server

This section contains options used when in server mode it will begin with `[server]`

On Wayland servers using InputCapture version 1, clipboard sharing falls back
to `wl-copy` and `wl-paste` from the `wl-clipboard` package when built with
libportal clipboard support. Enable clipboard sharing normally. Local content
is read when moving to another computer; remote content is published when
returning. The fallback supports UTF-8 text and PNG images, choosing PNG when
publishing a selection that also contains text. Transfers respect the configured
size limit and run in a separate worker with a timeout, so a stalled clipboard
owner does not block keyboard or mouse forwarding.

|Option              |    Valid Values   |Description|
|:-------------------|:-----------------:|:-----------|
| clipboardSize      | int > 0           | Deskflow will send a maximum of `N` megabytes of clipboard data to another computer when the mouse transitions to that computer. Clients use the server's value for both sending and receiving.|
| defaultLockToComputerState| `true` or `false` | When this is true the cursor is locked to the new computer when switching (default: false)|
| disableLockToComputer| `true` or `false` | If false pressing scroll lock will toggle your cursor to be locked to current computer. (default: false) |
| enableClipboard    | `true` or `false` | When `true` the clipboard will be shared with all clients If set to ''true'' then clipboard shared and the ''clipboardSharingSize'' setting will be used. If set to false, then clipboard sharing will be disabled and the the ''clipboardSharingSize'' setting will be ignored.|
| enableHeartbeat    | `true` or `false` | Send a heartbeat to connected clients; this has been replaced by internal keep alive (default: false)|
| enableSwitchDelay  | `true` or `false` | Switching will be delayed by the set value (default: false)|
| enableSwitchDoubleTap  | `true` or `false` | Enables the doubletap to switch method (default: false)|
| externalConfig     | `true` or `false` | When true use the external config path |
| externalConfigFile | Filepath          | Path the server config file if it does not exist the GUI will it generated based on the `internalConfig` section.|
| gridHeight         | int               | Height of the server's intenal grid used for the computer layout (default: 3)|
| gridWidth          | int               | Width of the server's intenal grid used for the computer layout (default: 5) |
| heartbeat          | int               | The server will expect each client to send a message no less than every `N` milliseconds. If no message arrives from a client within `3N` seconds the server forces that client to disconnect. If deskflow fails to detect clients disconnecting while the server is sleeping or vice versa, try using this option.|
| protocol           | `barrier` or `synergy` | The protocol to use when saying hello to clients. Can be set to barrier or synergy. If not set barrier is used as the default |
|relativeMouseMoves  | `true` or `false` | If set to ''true'' then secondary computers move the mouse using relative rather than absolute mouse moves when and only when the cursor is locked to the computer (by ''Scroll Lock'' or a configured hot key). This is intended to make Deskflow work better with certain games. If set to ''false'' or not set then all mouse moves are absolute.|
| switchDelay        | int               | Deskflow won't switch computers when the mouse reaches edge of a computer unless it stays on the edge for `N` milliseconds. This helps prevent unintentional switching when working near an edge. (default: 250)|
| switchDoubleTap    | int               | Deskflow won't switch computers when the mouse reaches the edge of a computer unless it's moved away from the edge and then back to the edge within `N` milliseconds. With the option you have to quickly tap the edge twice to switch. This helps prevent unintentional switching when working near the edge.|
|win32KeepForeground | `true` or `false` | If set to ''true'' (the default), Deskflow will grab the foreground focus on a Windows server (thereby putting all other windows in the background) upon switching to a client. If set to ''false'', it will leave the currently foreground window in the foreground. Deskflow grabs the focus to avoid issues with other apps interfering with Deskflow's ability to read the hardware inputs. |

 - You can use both the ''switchDelay'' and ''switchDoubleTap'' options at the same time. Deskflow will switch when either requirement is satisfied.

### Computer Settings

Each computer will have a section where its configuration will be stored, if the computer was named "foo" the section will be named `[computer_foo]`

|Option                  |    Valid Values   |Description|
|:-----------------------|:-----------------:|:-----------|
| aliases                | Comma separated list of hostnames | Names here will be used as alternatives for the computer. Names must be valid hostnames. |
| name                   | Valid hostname    | The name of the client. |
| halfDuplexCapsLock     | `true` or `false` | This computer has a ''Caps Lock'' key that doesn't report a press and a release event when the user presses it but instead reports a press event when it's turned on and a release event when it's turned off. If ''Caps Lock'' acts strangely on all computers then you may need to set this option to true on the server. If it acts strangely on one computer then that computer may need the option set to true.|
| halfDuplexNumLock      | `true` or `false` | This computer has a ''Num Lock'' key that doesn't report a press and a release event when the user presses it but instead reports a press event when it's turned on and a release event when it's turned off. If ''Num Lock'' acts strangely on all computers then you may need to set this option to true on the server. If it acts strangely on one computer then that computer may need the option set to true.|
| halfDuplexScrollLock   | `true` or `false` | This computer has a ''Scroll Lock'' key that doesn't report a press and a release event when the user presses it but instead reports a press event when it's turned on and a release event when it's turned off. If ''Scroll Lock'' acts strangely on all computers then you may need to set this option to true on the server. If it acts strangely on one computer then that computer may need the option set to true.|
| modifierAlt             | `shift` `ctrl` `alt` `altgr` `meta` `super` or  `none` | Map the server's alt modifer to different key on this computer|
| modifierAltGr           | `shift` `ctrl` `alt` `altgr` `meta` `super` or  `none` | Map the server's altgr modifer to different key on this computer|
| modifierCtrl            | `shift` `ctrl` `alt` `altgr` `meta` `super` or  `none` | Map the server's ctrl modifer to different key on this computer|
| modifierMeta            | `shift` `ctrl` `alt` `altgr` `meta` `super` or  `none` | Map the server's meta modifer to different key on this computer|
| modifierShift           | `shift` `ctrl` `alt` `altgr` `meta` `super` or  `none` | Map the server's shift modifer to different key on this computer|
| modifierSuper           | `shift` `ctrl` `alt` `altgr` `meta` `super` or  `none` | Map the server's super modifer to different key on thist computer|
| switchCornerSize       | integer (N)       | Sets the size of all corners in pixels. The cursor must be within `N` pixels of the corner to be considered to be in the corner.|
| switchCornerTopLeft    | `true` or `false` | Deskflow won't switch computers when the mouse reaches the edge of the computer if it includes this corner. The size of the corner is given by the `switchCornerSize` option. |
| switchCornerTopRight   | `true` or `false` | Deskflow won't switch computers when the mouse reaches the edge of the computer if it includes this corner. The size of the corner is given by the `switchCornerSize` option. |
| switchCornerBottomLeft | `true` or `false` | Deskflow won't switch computers when the mouse reaches the edge of the computer if it includes this corner. The size of the corner is given by the `switchCornerSize` option. |
| switchCornerBottomRight| `true` or `false` | Deskflow won't switch computers when the mouse reaches the edge of the computer if it includes this corner. The size of the corner is given by the `switchCornerSize` option. |
| xtestIsXineramaUnaware | `true` or `false` | This option works around a bug in the XTest extension when used in combination with Xinerama. It affects X11 clients only. Not all versions of the XTest extension are aware of the Xinerama extension. As a result, they do not move the mouse correctly when using multiple Xinerama screens. This option is currently ''true'' by default. If you know your XTest extension is Xinerama aware then set this option to ''false''.|
| weakX11Focus           | `true` or `false` | This X11 Client will not gain focus until an input event has been processed its any of its screens |

### InternalConfig

This section contains options used when in server mode it will begin with `[internalConfig]`

src/lib/common/Settings.cpp
/*
 * Deskflow -- mouse and keyboard sharing utility
 * SPDX-FileCopyrightText: (C) 2025 - 2026 Chris Rizzitello <sithlord48@gmail.com>
 * SPDX-License-Identifier: GPL-2.0-only WITH LicenseRef-OpenSSL-Exception
 */

#include "Settings.h"

#include "LogLevel.h"
#include "NetworkProtocol.h"
#include "UrlConstants.h"
#include "common/KeyboardModifier.h"

#include <QCoreApplication>
#include <QFile>
#include <QFileSystemWatcher>
#include <QRect>
#include <QRegularExpression>
#include <QStandardPaths>

Settings *Settings::instance()
{
  static Settings m;
  return &m;
}

void Settings::setSettingsFile(const QString &settingsFile)
{
  if (Settings::settingsFile() == settingsFile) {
    qDebug("settings file already set, skipping");
    return;
  }

  if (instance()->m_settings)
    instance()->m_settings->deleteLater();

  instance()->m_settingsWatcher->removePath(instance()->m_settings->fileName());
  instance()->m_settings = new QSettings(settingsFile, QSettings::IniFormat, instance());
  instance()->m_settingsWatcher->addPath(settingsFile);
  instance()->m_settingsProxy->load(settingsFile);
  qInfo().noquote() << "settings file changed:" << instance()->m_settings->fileName();

  instance()->upgradeSettings();
  instance()->cleanSettings();
  instance()->cleanStateSettings();
  instance()->setupComputerName();
  instance()->checkIfSettingsWritableChange();
}

void Settings::setStateFile(const QString &stateFile)
{
  if (instance()->m_stateSettings->fileName() == stateFile) {
    qDebug("state file already set, skipping");
    return;
  }

  if (instance()->m_stateSettings)
    instance()->m_stateSettings->deleteLater();

  instance()->m_stateSettings = new QSettings(stateFile, QSettings::IniFormat, instance());
  qInfo().noquote() << "state file changed:" << instance()->m_stateSettings->fileName();
}

void Settings::clearState()
{
  instance()->m_stateSettings->clear();
  instance()->m_stateSettings->sync();
}

Settings::Settings(QObject *parent) : QObject(parent), m_settingsWatcher{new QFileSystemWatcher(this)}
{
  QString fileToLoad;
#ifdef Q_OS_WIN
  const auto portableFile = portableSettingsFile();
  qDebug().noquote() << "checking for portable settings file at:" << portableFile;
  if (QFile(portableFile).exists())
    fileToLoad = portableFile;
#else
  if (const auto xdgConfigHome = qEnvironmentVariable("XDG_CONFIG_HOME"); !xdgConfigHome.isEmpty())
    fileToLoad = QStringLiteral("%1/%2/%2.conf").arg(xdgConfigHome, kAppName);
#endif
  else if (QFile(UserSettingFile).exists())
    fileToLoad = UserSettingFile;
  else if (QFile(SystemSettingFile).exists())
    fileToLoad = SystemSettingFile;
  else
    fileToLoad = UserSettingFile;

  m_settings = new QSettings(fileToLoad, QSettings::IniFormat, this);
  m_settingsWritable = m_settings->isWritable();
  m_settingsWatcher->addPath(m_settings->fileName());
  connect(m_settingsWatcher, &QFileSystemWatcher::fileChanged, this, &Settings::checkIfSettingsWritableChange);
  m_settingsProxy = std::make_shared<QSettingsProxy>();
  m_settingsProxy->load(fileToLoad);
  qInfo().noquote() << "initial settings file:" << m_settings->fileName();

  const auto xdgStateHome = qEnvironmentVariable("XDG_STATE_HOME");
  const auto stateBase = !xdgStateHome.isEmpty()
                             ? xdgStateHome
                             : QStandardPaths::standardLocations(QStandardPaths::GenericStateLocation).at(0);
  const auto stateFile = QStringLiteral("%1/%2.state").arg(stateBase, kAppName);

  m_stateSettings = new QSettings(stateFile, QSettings::IniFormat, this);

  upgradeSettings();
  cleanSettings();
  cleanStateSettings();
  setupComputerName();
}

void Settings::upgradeSettings()
{
  if (const auto logValue = m_settings->value(Settings::Log::Level).toString();
      !LogLevel::logLevelOptions().contains(logValue, Qt::CaseInsensitive))
    m_settings->setValue(Settings::Log::Level, defaultValue(Settings::Log::Level));
  const auto screenKeys = m_settings->allKeys().filter("screen_");
  for (const auto &key : screenKeys) {
    auto newKey = key;
    newKey.replace("screen_", "computer_");
    m_settings->setValue(newKey, m_settings->value(key));
  }
  for (const auto [oldKey, newKey] : m_upgradedMap.asKeyValueRange()) {
    if (m_settings->contains(newKey) || !m_settings->contains(oldKey))
      continue;
    if (oldKey == InternalConfig::Protocol) {
      m_settings->setValue(newKey, networkProtocolToOption(NetworkProtocol(m_settings->value(oldKey).toInt())));
    } else if (oldKey == InternalConfig::ClipboardSharingSize) {
      m_settings->setValue(newKey, m_settings->value(oldKey).toUInt() / 1024);
    } else {
      m_settings->setValue(newKey, m_settings->value(oldKey));
    }
  }

  // move to computers

  m_settings->beginGroup("internalConfig");
  const int screenArraySize = m_settings->beginReadArray("screens");
  m_settings->endArray();
  m_settings->endGroup();

  for (int i = 1; i <= screenArraySize; i++) {
    m_settings->beginGroup("internalConfig");
    m_settings->beginReadArray("screens");
    m_settings->setArrayIndex(i);
    const auto name = m_settings->value("name", {}).toString();

    if (name.isEmpty()) {
      m_settings->endArray();
      m_settings->endGroup();
      continue;
    }

    int switchCornerSize = 0;
    if (m_settings->contains("switchCornerSize"))
      switchCornerSize = m_settings->value("switchCornerSize", 0).toInt();

    QList<bool> fixTemp;
    if (m_settings->contains("fixArray/size")) {
      int fixArraySize = m_settings->beginReadArray("fixArray");
      for (int j = 0; j < fixArraySize; j++) {
        m_settings->setArrayIndex(j);
        fixTemp.append(m_settings->value("fix").toBool());
      }
      m_settings->endArray();
    }

    QList<bool> cornerTemp;
    if (m_settings->contains("switchCornerArray/size")) {
      int cornerArraySize = m_settings->beginReadArray("switchCornerArray");
      for (int j = 0; j < cornerArraySize; j++) {
        m_settings->setArrayIndex(j);
        cornerTemp.append(m_settings->value("switchCorner").toBool());
      }
      m_settings->endArray();
    }

    QList<QString> modifierTemp;
    if (m_settings->contains("modifierArray/size")) {
      int modifierArraySize = m_settings->beginReadArray("modifierArray");
      for (int j = 0; j < modifierArraySize; j++) {
        m_settings->setArrayIndex(j);
        auto modifier = m_settings->value("modifier").toInt();
        modifierTemp.append(
            modifier == -1 ? valueToKeyboardModifierOption(j) : valueToKeyboardModifierOption(modifier)
        );
      }
      m_settings->endArray();
    }

    QStringList aliases;
    if (m_settings->contains("aliasArray/size")) {
      int aliasList = m_settings->beginReadArray("aliasArray");
      for (int j = 0; j < aliasList; j++) {
        m_settings->setArrayIndex(j);
        aliases.append(m_settings->value("alias").toString());
      }
      m_settings->endArray();
      aliases.removeDuplicates();
    }
    m_settings->endArray();
    m_settings->endGroup();

    // Write Settings
    if (!m_settings->contains(Settings::Computer::Name.arg(name)))
      m_settings->setValue(Settings::Computer::Name.arg(name), name);
    if (!m_settings->contains(Settings::Computer::SwitchCornerSize.arg(name)))
      m_settings->setValue(Settings::Computer::SwitchCornerSize.arg(name), switchCornerSize);
    if (!m_settings->contains(Settings::Computer::HalfDuplexCapsLock.arg(name)))
      m_settings->setValue(Settings::Computer::HalfDuplexCapsLock.arg(name), fixTemp.at(0));
    if (!m_settings->contains(Settings::Computer::HalfDuplexNumLock.arg(name)))
      m_settings->setValue(Settings::Computer::HalfDuplexNumLock.arg(name), fixTemp.at(1));
    if (!m_settings->contains(Settings::Computer::HalfDuplexScrollLock.arg(name)))
      m_settings->setValue(Settings::Computer::HalfDuplexScrollLock.arg(name), fixTemp.at(2));
    if (!m_settings->contains(Settings::Computer::XtestIsXineramaUnaware.arg(name)))
      m_settings->setValue(Settings::Computer::XtestIsXineramaUnaware.arg(name), fixTemp.at(3));
    if (!m_settings->contains(Settings::Computer::SwitchCornerTopLeft.arg(name)))
      m_settings->setValue(Settings::Computer::SwitchCornerTopLeft.arg(name), cornerTemp.at(0));
    if (!m_settings->contains(Settings::Computer::SwitchCornerTopRight.arg(name)))
      m_settings->setValue(Settings::Computer::SwitchCornerTopRight.arg(name), cornerTemp.at(1));
    if (!m_settings->contains(Settings::Computer::SwitchCornerBottomLeft.arg(name)))
      m_settings->setValue(Settings::Computer::SwitchCornerBottomLeft.arg(name), cornerTemp.at(2));
    if (!m_settings->contains(Settings::Computer::SwitchCornerBottomRight.arg(name)))
      m_settings->setValue(Settings::Computer::SwitchCornerBottomRight.arg(name), cornerTemp.at(3));
    if (!m_settings->contains(Settings::Computer::ModifierShift.arg(name)))
      m_settings->setValue(Settings::Computer::ModifierShift.arg(name), modifierTemp.at(0));
    if (!m_settings->contains(Settings::Computer::ModifierCtrl.arg(name)))
      m_settings->setValue(Settings::Computer::ModifierCtrl.arg(name), modifierTemp.at(1));
    if (!m_settings->contains(Settings::Computer::ModifierAlt.arg(name)))
      m_settings->setValue(Settings::Computer::ModifierAlt.arg(name), modifierTemp.at(2));
    if (!m_settings->contains(Settings::Computer::ModifierMeta.arg(name)))
      m_settings->setValue(Settings::Computer::ModifierMeta.arg(name), modifierTemp.at(3));
    if (!m_settings->contains(Settings::Computer::ModifierSuper.arg(name)))
      m_settings->setValue(Settings::Computer::ModifierSuper.arg(name), modifierTemp.at(4));
    if (!m_settings->contains(Settings::Computer::ModifierAltGr.arg(name)))
      m_settings->setValue(Settings::Computer::ModifierAltGr.arg(name), modifierTemp.at(5));
    if (!m_settings->contains(Settings::Computer::Aliases.arg(name))) {
      m_settings->setValue(Settings::Computer::Aliases.arg(name), aliases);
    }
  }
}

void Settings::cleanSettings()
{
  const QStringList keys = m_settings->allKeys();
  for (const QString &key : keys) {
    if (m_oldServerConfigKeys.contains(key) || key.startsWith("screen_"))
      m_settings->remove(key);
    if (key.startsWith(QStringLiteral("internalConfig")))
      continue;
    if (const auto group = key.mid(0, key.indexOf('/')); !m_validKeys.contains(key) && m_validGroup.contains(group))
      m_settings->remove(key);
    if (!m_settings->value(key).isValid())
      m_settings->remove(key);
  }
}

void Settings::cleanStateSettings()
{
  const QStringList keys = m_stateKeys;
  for (const QString &key : keys) {
    if (!m_stateKeys.contains(key))
      m_stateSettings->remove(key);
    if (m_stateSettings->value(key).toString().isEmpty() && !m_stateSettings->value(key).toRect().isValid())
      m_stateSettings->remove(key);
  }
}

void Settings::setupComputerName()
{
  if (m_settings->value(Settings::Core::ComputerName).toString().isEmpty())
    m_settings->setValue(Settings::Core::ComputerName, cleanComputerName(QSysInfo::machineHostName()));
}

QString Settings::cleanComputerName(const QString &name)
{
  static const auto hyphen = QStringLiteral("-");
  static const auto space = QStringLiteral(" ");
  static const auto underscore = QStringLiteral("_");
  static const auto period = QStringLiteral(".");
  static const auto nameRegex = QRegularExpression(QStringLiteral("[^\\w\\-\\.]"));

  QString cleanName = name.simplified();
  cleanName.replace(space, underscore);
  cleanName.replace(nameRegex, {});
  while (cleanName.startsWith(hyphen) || cleanName.startsWith(underscore) || cleanName.startsWith(period))
    cleanName.removeFirst();
  while (cleanName.endsWith(hyphen) || cleanName.endsWith(underscore) || cleanName.endsWith(period))
    cleanName.removeLast();
  if (cleanName.length() > 255) {
    cleanName.truncate(255);
    cleanName = cleanComputerName(cleanName);
  }
  return cleanName;
}

void Settings::checkIfSettingsWritableChange()
{
  if (!instance()->m_settingsWatcher->files().contains(Settings::settingsFile()))
    m_settingsWatcher->addPath(Settings::settingsFile());

  bool writable = Settings::isWritable();
  if (writable == m_settingsWritable)
    return;
  qDebug() << QString("Setting are now %1").arg(writable ? "writable" : "readonly");
  instance()->m_settingsWritable = writable;
  Q_EMIT instance()->settingsWritableChanged(writable);
}

QVariant Settings::defaultValue(const QString &key)
{
  if (key.startsWith(QStringLiteral("computer_"))) {
    const auto computerKey = key.mid(key.indexOf('/') + 1);
    return computerDefaults(computerKey);
  }

  if (m_defaultFalseValues.contains(key))
    return false;

  if (m_defaultTrueValues.contains(key))
    return true;

  if (key == Security::Certificate)
    return QStringLiteral("%1/%2.pem").arg(Settings::tlsDir(), kAppId);

  if (key == Security::KeySize)
    return 2048;

  if (key == Log::File)
    return QStringLiteral("%1/%2.log").arg(QDir::homePath(), kAppId);

  if (key == Log::Level)
    return QVariant::fromValue(LogLevel::Level::Info).toString();

  if (key == Daemon::Elevate)
    return !Settings::isPortableMode();

  if (key == Gui::UpdateCheckUrl)
    return kUrlUpdateCheck;

  if (key == Server::ExternalConfigFile)
    return QStringLiteral("%1/%2-server.conf").arg(Settings::settingsPath(), kAppId);

  if (key == Core::Port)
    return 24800;

  if (key == Core::ProcessMode) {
#ifdef Q_OS_WIN
    if (!Settings::isPortableMode())
      return Settings::ProcessMode::Service;
#endif
    return Settings::ProcessMode::Desktop;
  }

  if (key == Daemon::LogFile)
    return QStringLiteral("%1/%2-daemon.log").arg(Settings::settingsPath(), kAppId);

  if (key == Client::YScrollScale || key == Client::XScrollScale)
    return 1.0;

  if (key == Server::Protocol)
    return networkProtocolToOption(NetworkProtocol::Barrier);

  if (key == Server::GridWidth)
    return kServerGridWidth;

  if (key == Server::GridHeight)
    return kServerGridHeight;

  if (key == Server::Heartbeat)
    return 5000;

  if (key == Server::SwitchDelay || key == Server::SwitchDoubleTap)
    return 250;

  if (key == Server::ClipboardSize)
    return 3; // 3 MiB

  if (key == Server::SwapControlSuperScreens)
    return QStringList();

  return QVariant();
}

QSettingsProxy &Settings::proxy()
{
  return *instance()->m_settingsProxy;
}

NetworkProtocol Settings::networkProtocol()
{
  return networkProtocolFromString(Settings::value(Server::Protocol).toString());
}

void Settings::save(bool emitSaving)
{
  if (!Settings::isWritable())
    qWarning().noquote() << "settings not saved, file is read-only:" << Settings::settingsFile();
  if (emitSaving)
    Q_EMIT instance()->serverSettingsChanged();
  instance()->m_settings->sync();
  instance()->m_stateSettings->sync();
}

QStringList Settings::validKeys()
{
  return Settings::m_validKeys;
}

QString Settings::serverConfigFile()
{
  bool useExt = value(Server::ExternalConfig).toBool();
  return useExt ? value(Server::ExternalConfigFile).toString() : defaultValue(Server::ExternalConfigFile).toString();
}

bool Settings::isServerConfigFileReadable()
{
  auto file = QFile(serverConfigFile());
  return file.open(QFile::ReadOnly);
}

bool Settings::isWritable()
{
  return instance()->m_settings->isWritable();
}

bool Settings::isPortableMode()
{
  // Enable portable mode only if the portable settings file exists in the expected location.
  return QFile(portableSettingsFile()).exists();
}

QString Settings::settingsFile()
{
  return QFileInfo(instance()->m_settings->fileName()).absoluteFilePath();
}

QString Settings::settingsPath()
{
#ifdef Q_OS_WIN
  if (!isPortableMode())
    return SystemDir;
#endif

  return QFileInfo(instance()->m_settings->fileName()).absolutePath();
}

QString Settings::tlsDir()
{
  return QStringLiteral("%1/tls").arg(instance()->settingsPath());
}

QString Settings::tlsTrustedServersDb()
{
  return QFileInfo(QStringLiteral("%1/trusted-servers").arg(instance()->tlsDir())).absoluteFilePath();
}

QString Settings::tlsTrustedClientsDb()
{
  return QFileInfo(QStringLiteral("%1/trusted-clients").arg(instance()->tlsDir())).absoluteFilePath();
}

QString Settings::logLevelText()
{
  return Settings::value(Log::Level).toString();
}

void Settings::setValue(const QString &key, const QVariant &value)
{
  const bool useState = Settings::m_stateKeys.contains(key) && !instance()->isPortableMode();
  auto settings = useState ? instance()->m_stateSettings : instance()->m_settings;

  if (settings->value(key) == value)
    return;

  if (!value.isValid())
    settings->remove(key);
  else {
    if (key == Settings::Core::ComputerName)
      settings->setValue(key, cleanComputerName(value.toString()));
    else
      settings->setValue(key, value);
  }

  settings->sync();
  Q_EMIT instance()->settingsChanged(key);
}

QVariant Settings::value(const QString &key)
{
  const bool useState = Settings::m_stateKeys.contains(key) && !instance()->isPortableMode();
  auto settings = useState ? instance()->m_stateSettings : instance()->m_settings;
  return settings->value(key, defaultValue(key));
}

void Settings::restoreDefaultSettings()
{
  for (const auto &key : m_validKeys) {
    instance()->setValue(key, defaultValue(key));
  }
}

QString Settings::portableSettingsFile()
{
  static const auto filename =
      QStringLiteral("%1/settings/%2.conf").arg(QCoreApplication::applicationDirPath(), kAppName);
  return QFileInfo(filename).absoluteFilePath();
}

void Settings::removeUnknownComputers(const QStringList &knownComputers)
{
  const QStringList knownGroups = instance()->m_settings->childGroups();
  for (const auto &group : knownGroups) {
    if (m_validGroup.contains(group) || knownComputers.contains(group))
      continue;
    instance()->m_settings->remove(group);
  }
}

QVariant Settings::computerDefaults(const QString &key)
{
  if (key.contains(QStringLiteral("halfDuplex")) || key.contains("xtestIsXineramaUnaware") || key.contains("X11") ||
      key.contains("Left") || key.contains("Right"))
    return false;
  if (key.endsWith("switchCornerSize"))
    return 0;
  if (key == "modifierAlt")
    return keyboardModifierToOption(KeyboardModifier::Alt);
  if (key == "modifierAltGr")
    return keyboardModifierToOption(KeyboardModifier::AltGr);
  if (key == "modifierCtrl")
    return keyboardModifierToOption(KeyboardModifier::Ctrl);
  if (key == "modifierMeta")
    return keyboardModifierToOption(KeyboardModifier::Meta);
  if (key == "modifierShift")
    return keyboardModifierToOption(KeyboardModifier::Shift);
  if (key == "modifierSuper")
    return keyboardModifierToOption(KeyboardModifier::Super);
  return QVariant();
}

QStringList Settings::knownComputers()
{
  const QStringList knownGroups = instance()->m_settings->childGroups();
  QStringList computers;
  for (const auto &group : knownGroups) {
    if (group.startsWith("computer_")) {
      computers.append(Settings::value(QStringLiteral("%1/name").arg(group)).toString());
    }
  }
  return computers;
}
